browser/ui/webui/ai_chat/ai_chat_ui.cc

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/ai_chat/ai_chat_ui.h"
+
+#include <utility>
+
+#include "brave/browser/ai_chat/common/constants.h"
+#include "brave/browser/ai_chat/common/pref_names.h"
+#include "brave/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.h"
+#include "brave/browser/ui/webui/brave_webui_source.h"
+#include "brave/components/ai_chat_ui/resources/page/grit/ai_chat_ui_generated_map.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "brave/components/l10n/common/localization_util.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "components/grit/brave_components_resources.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "content/public/common/url_constants.h"
+#include "services/network/public/mojom/content_security_policy.mojom.h"
+
+AIChatUI::AIChatUI(content::WebUI* web_ui)
+    : ui::UntrustedWebUIController(web_ui),
+      profile_(Profile::FromWebUI(web_ui)) {
+  DCHECK(profile_);
+  browser_ = chrome::FindLastActiveWithProfile(profile_);
```

Comment:
Nice catch. I went with FindBrowserWithWebContents approach

Reply:
Oops, FindBrowserWithWebContents crashes

```
Received signal 11 SEGV_ACCERR 000000000230
0   libbase.dylib                       0x000000010326fd5c base::debug::CollectStackTrace(void**, unsigned long) + 28
1   libbase.dylib                       0x00000001032582f8 base::debug::StackTrace::StackTrace() + 24
2   libbase.dylib                       0x000000010326fc54 base::debug::(anonymous namespace)::StackDumpSignalHandler(int, __siginfo*, void*) + 1204
3   libsystem_platform.dylib            0x000000019af22a84 _sigtramp + 56
4   libchrome_dll.dylib                 0x000000010dcb6f24 content::InterfaceRegistrationHelper<AIChatUI>& content::InterfaceRegistrationHelper<AIChatUI>::Add<ai_chat::mojom::PageHandler>()::'lambda'(mojo::BinderMapWithContext<content::WebUIController*>*)::operator()(mojo::BinderMapWithContext<content::WebUIController*>*) const::'lambda'(content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>)::operator()(content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>) const + 104
5   libchrome_dll.dylib                 0x000000010dcb6f24 content::InterfaceRegistrationHelper<AIChatUI>& content::InterfaceRegistrationHelper<AIChatUI>::Add<ai_chat::mojom::PageHandler>()::'lambda'(mojo::BinderMapWithContext<content::WebUIController*>*)::operator()(mojo::BinderMapWithContext<content::WebUIController*>*) const::'lambda'(content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>)::operator()(content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>) const + 104
6   libchrome_dll.dylib                 0x000000010dcb6ea4 base::internal::Invoker<base::internal::BindState<content::InterfaceRegistrationHelper<AIChatUI>& content::InterfaceRegistrationHelper<AIChatUI>::Add<ai_chat::mojom::PageHandler>()::'lambda'(mojo::BinderMapWithContext<content::WebUIController*>*)::operator()(mojo::BinderMapWithContext<content::WebUIController*>*) const::'lambda'(content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>)>, void (content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>)>::Run(base::internal::BindStateBase*, content::WebUIController*, mojo::PendingReceiver<ai_chat::mojom::PageHandler>&&) + 52
7   libchrome_dll.dylib                 0x000000010ca5c618 base::RepeatingCallback<void (CoreAccountId const&, std::Cr::basic_string<char, std::Cr::char_traits<char>, std::Cr::allocator<char>> const&)>::Run(CoreAccountId const&, std::Cr::basic_string<char, std::Cr::char_traits<char>, std::Cr::allocator<char>> const&) const & + 92
8   libchrome_dll.dylib                 0x000000010cb8b1a4 void mojo::internal::BinderContextTraits<content::RenderFrameHost*>::BindGenericReceiver<extensions::KeepAlive>(base::RepeatingCallback<void (content::RenderFrameHost*, mojo::PendingReceiver<extensions::KeepAlive>)> const&, content::RenderFrameHost*, mojo::ScopedHandleBase<mojo::MessagePipeHandle>) + 80
9   libchrome_dll.dylib                 0x000000010cb8b1e4 base::internal::Invoker<base::internal::BindState<void (*)(base::RepeatingCallback<void (content::RenderFrameHost*, mojo::PendingReceiver<extensions::KeepAlive>)> const&, content::RenderFrameHost*, mojo::ScopedHandleBase<mojo::MessagePipeHandle>), base::RepeatingCallback<void (content::RenderFrameHost*, mojo::PendingReceiver<extensions::KeepAlive>)>>, void (content::RenderFrameHost*, mojo::ScopedHandleBase<mojo::MessagePipeHandle>)>::Run(base::internal::BindStateBase*, content::RenderFrameHost*, mojo::ScopedHandleBase<mojo::MessagePipeHandle>&&) + 40
10  libcontent.dylib                    0x0000000107153a44 base::RepeatingCallback<void (url::Origin const&, blink::mojom::IdpSigninStatus)>::Run(url::Origin const&, blink::mojom::IdpSigninStatus) const & + 92
11  libcontent.dylib                    0x00000001077ee9a4 mojo::internal::GenericCallbackBinderWithContext<content::WebUIController*>::BindInterface(content::WebUIController*, mojo::ScopedHandleBase<mojo::MessagePipeHandle>) + 160
12  libcontent.dylib                    0x00000001077ee828 mojo::BinderMapWithContext<content::WebUIController*>::TryBind(content::WebUIController*, mojo::GenericPendingReceiver*) + 164
13  libcontent.dylib                    0x00000001077ee6c4 content::PerWebUIBrowserInterfaceBroker::GetInterface(mojo::GenericPendingReceiver) + 252
14  libblink_common.dylib               0x00000001172b471c blink::mojom::BrowserInterfaceBrokerStubDispatch::Accept(blink::mojom::BrowserInterfaceBroker*, mojo::Message*) + 228
```

Reply:
There's another variant which I see is: `chrome::FindBrowserWithProfile`

Reply:
Well, this is technically not correct - since still, the Browser that we're looking for might not have been the most recently active. I'm sure it's going to be fine 99.9% of the time, but we should really do this the right way if you have a moment - the constructor (ChatUISidePanelCoordinator) does know the correct Browser to use and can call an `Initialize(Browser* browser)` function on AIChatUI

Reply:
Fixed. I have done this via side panel's AIChat helper.

You'll see the header in `b/b/ui/side_panel/ai_chat` and it's source in `b/b/ui/views/side_panel/ai_chat` this way I can include the helper in bbui webui without any layer violation.